src/yz_app.erl

```diff
@@ -29,25 +29,76 @@
 -define(QUERY_SERVICES, [{yz_pb_search, 27, 28}]).
 -define(ADMIN_SERVICES, [{yz_pb_admin, 54, 60}]).
 
+%% NOTE: These default values are duplicated in yokozuna.schema.
+-define(YZ_CONFIG_IBROWSE_MAX_SESSIONS_DEFAULT, 100).
+-define(YZ_CONFIG_IBROWSE_MAX_PIPELINE_SIZE_DEFAULT, 1).
+
+components() ->
+    [index, search].
+
 %%%===================================================================
 %%% Callbacks
 %%%===================================================================
 
 start(_StartType, _StartArgs) ->
+
+    %% Disable indexing/searching from KV until properly started,
+    %% otherwise restarting under load generates large numbers
+    %% of failures in yz_kv:index/3.
+    disable_components(),
+
+    %% Ensure that the KV service has fully loaded.
+    riak_core:wait_for_service(riak_kv),
+
+
     initialize_atoms(),
     Enabled = ?YZ_ENABLED,
     case yz_sup:start_link(Enabled) of
         {ok, Pid} ->
             _ = application:set_env(ibrowse, inactivity_timeout, 600000),
             maybe_setup(Enabled),
+
+            %% Now everything is started, permit usage by KV/query
+            enable_components(),
             {ok, Pid};
         Error ->
             Error
     end.
 
+%% @doc Prepare to stop - called before the supervisor tree is shutdown
+prep_stop(State) ->
+    try %% wrap with a try/catch - application carries on regardless,
+        %% no error message or logging about the failure otherwise.
+        lager:info("Stopping application yokozuna.", []),
+        riak_api_pb_service:deregister(?QUERY_SERVICES),
+        riak_api_pb_service:deregister(?ADMIN_SERVICES),
```

Comment:
I didn't get a chance to review, but from my [initial intentions](https://github.com/basho/yokozuna/issues/581), i think we should try to if possible.